Update weight init to general use cases

KushajveerSingh opened the pull request:
Fixes #18182 

### Usage example
```python
l = nn.Linear(2,3)
l.bias.data
# tensor([0., 0., 0.])

with nn.init.init_version('1.6.1') as version:
    l = nn.Linear(2,3)
l.bias.data
# tensor([-0.6192, -0.2009, -0.6798])
```

cc @t-vi @apaszke @gchanan @soumith @colesbury

KushajveerSingh committed:
New context manager for weight init

A new context manager `nn.init_version` is introduced that will allow us
to introduce new breaking initialization schemes in pytorch, with
backward compataibility.

Currenlty, it is only implemented for `nn.Linear`.

KushajveerSingh committed:
remove packaging dependency

zhangguanheng66 commented:
cc@ezyang, seems an old issue. Are we still pursuing PR to fix this?

ezyang commented:
If the version mechanism is implemented correctly it won't be BC breaking. The weight init updates are long overdue. cc @soumith 

@KushajveerSingh what's your game plan, and do you have reviewers signed up to work with you on this?

KushajveerSingh commented:
Had to reset my laptop, so could not work on it. We can make the version logic opposite. If someone wants to use newer initialization scheme, then they can use the context manager. This will not break all the pytorch repos till now. This also allows us to more freely add new initializations for people to try out without fear of breaking things. 

My plan is to make the initialization work for the general conv-relu case. As there are so many variations in the architectures, coming up with a common initialization is not possible.

Also, this will provide a way to add new initialization schemes in the future. 

I don't have reviewers signed up.

KushajveerSingh committed:
Make this a non-breaking change.

By default, the pytorch initialization used till 1.6.1 is used. This
makes this a non-breaking change, and allows us to more freely add new
initialization schemes in pytorch.

t-vi commented:
Good to see this worked on, thank you @KushajveerSingh .

So while I think it's OK to make this non-breaking for now, I think that at some point we want to deprecate the current behaviour with an indication how to get "fixed" vs. "current" bahavior and then eventually default to current.

KushajveerSingh commented:
Made this a non-breaking change. Now we do not have to worry about breaking all the pytorch code till today.

Usage example (suppose this PR is accepted in 1.7.0)
```python
# Same as till 1.6.1
l = nn.Linear(2,3)

# If we want a specific version initialization
with nn.init.init_version('1.7.0') as version:
    l = nn.Linear(2,3)

# If we want to use the latest initialization always
with nn.init.init_version(use_master=True) as version:
    l = nn.Linear(2,3)
```

t-vi commented:
@KushajveerSingh I would imagine that the use pattern would be more commonly

```python
import torch
torch.nn.init.set_init_version('1.7.1')

# everything else
```
or somesuch, as likely the user either wants "the latest and greatest" or "whatever it was when she wrote the code".

KushajveerSingh commented:
> @KushajveerSingh I would imagine that the use pattern would be more commonly
> 
> ```python
> import torch
> torch.nn.init.set_init_version('1.7.1')
> ```

This is possible. Everything in this code is controlled by `nn.init._init_version=(1,6,1)` variable. A user can change this variable as per their needs.

But changing this variable carelessly can lead to some problems (maybe in the future). So we worked on a context manager so that it will handle the changing of this value and then setting it back to its default. 

This was suggested by @elistevens in this [message](https://github.com/pytorch/pytorch/issues/18182#issuecomment-659699030).

KushajveerSingh commented:
Notes on initialization (Correct me if I am wrong here):
1. Normalization layers:- People generally use the default pytorch initialization today also (assuming it is not the last layer before reset block and other special cases like that). So no need to make changes here.
2. Transformer:- In the paper, Xavier init is used (same as pytorch). I asked this on **huggingface forums** [link](https://discuss.huggingface.co/t/state-of-the-art-technique-for-initializing-embedding-matrix/326) but was not able to come to a convincing conclusion.
3. RNNBase, Embedding:- I have no idea.
4. Conv Layers, Linear, Bilinear:- As most of the people use linear-relu combination, so I think it is a good idea to change this initialization to `kaiming_normal_(mode='fan_out')`.

KushajveerSingh commented:
Is there a need to add **tests** for this change? Currently, none of the pytorch tests is testing the new changes that I introduced.

t-vi commented:
I think @Kaixhin 's list from the issue should be reasonable (so normalization with 1 var, 0 std at start).
Also note that it's not just calling the right function, but also not providing funny parameters.

t-vi commented:
re testing: I think you want tests. compare std and mean to expectation with some fixed random seed (for each test) to be within some CI of the expectation.

KushajveerSingh commented:
> I think @Kaixhin 's list from the issue should be reasonable (so normalization with 1 var, 0 std at start).

I checked the list. It also misses the initialization for Embedding, RNN. Others are the same as I showed above.

> Also note that it's not just calling the right function, but also not providing funny parameters.

Initially, I did this by using the `packaging` package. I would do `packaging.version.parse` and if the result was of type `packaging.version.Version` then I would proceed, else raise an error. But I had to remove `packaging` as it was a new dependency for pytorch.

Now I am doing the checking as follows for `version`:
1. if tuple, then check all values are int
2, if string, then split at `.`. I can improve this by checking if the string has two dots and other values are 0-9.

> re testing: I think you want tests. compare std and mean to expectation with some fixed random seed (for each test) to be within some CI of the expectation.

I will work on this.

Currently, some type annotation tests are failing so working on those.

KushajveerSingh committed:
solve type annotation issue

KushajveerSingh committed:
Add checks for version in _parse_version

Now only versions of type Tuple(int,int,int) or 'major.minor.micor' can
be passed in init_version.

KushajveerSingh committed:
remove type annotations

KushajveerSingh committed:
Add test cases for nn.init.init_version

Four test cases are added, that do the following
1. For version < 1.7.0, check old init is used
2. For version = 1.7.0, check new init is used
3. Ensure _init_version remains same after using init_version
4. Ensure you cannot pass version > torch.__version__ in init_version

KushajveerSingh committed:
solve linter issue for long line

KushajveerSingh committed:
add init_version in bilinear, _ConvNd

KushajveerSingh commented:
New progress checklist
- [x] Linear
- [x] Bilinear
- [x] _ConvNd
- [ ] RNNBase
- [ ] Embedding, EmbeddingBag
- [ ] Transformer
- [ ] Make version list page
- [x] Add test cases

KushajveerSingh committed:
solve linter issues

KushajveerSingh commented:
For the documentation page, I am looking to something as shown below:
```
1.9.0
New initialization introduced

1.7.0 to 1.8.1
Initialization did not change, so group the versions together

< 1.7.0
List old initialization
```
One thing that I would prefer is to have all the initialization listings to be independent. If in 1.9.0, a change was only made to `Linear`, I would still want 1.9.0 to include the details of all initialization, something like this:
```python
1.9.0
* Breaking
    Linear changes (details ...)
* Same
    Copy the initialization from the previous version here (everything except Linear)
```
By doing this, we can just open the page and get details of all the initialization for a particular version in a single place.

Can someone comment on this? Is this a correct approach? How to write docs page for PyTorch?

KushajveerSingh committed:
import Union in conv.py

t-vi commented:
I'd not overthink the documentation. It'll be in the release notes and then it's OK. I think documenting the current init scheme along with the version it was introduced for each layer (similar to the current documentation) would be sufficient, at least once we default to "current".

KushajveerSingh commented:
@t-vi Ok, will add the documentation to the modules that I changed.

I think the bulk work has been done. We have a system that allows us to add new initialization schemes easily (we can change the default behavior to always use the latest initialization in the future if we want, will document the changes to be made for this). 

I searched for the remaining layers initializations and they are almost identical with the current pytorch initialization. So there is no need to change those for now.

elistevens commented on torch/nn/init.py:
```diff
@@ -525,3 +527,104 @@ def deprecated_init(*args, **kwargs):
+
+v = torch.__version__.split('.')
+# These can also be used as global flags
+_torch_version = (int(v[0]), int(v[1]), int(v[2][0]))
```
We'll have to be careful that there's not ever a version `x.y.12` since will `nt(v[2][0])` mishandle that. probably safe?

t-vi replied:
Thanks, @elistevens . I think it would be best to use the same re pattern as `torch.utils.cpp_extension.BUILT_FROM_SOURCE_VERSION_PATTERN`. That way we have an even number of breakages, hopefully `0`.

KushajveerSingh replied:
Changed the `_torch_version` logic with re search.

```python
def _get_torch_version():
    """Returns torch.__version__ as Tuple(int,int,int)"""
    GET_TORCH_VERSION = re.compile(r'\d+\.\d+\.\d+')
    v = GET_TORCH_VERSION.findall(torch.__version__)
    assert len(v) == 1, "GET_TORCH_VERSION failed to find torch version"
    v = v[0].split('.')
    return (int(v[0]), int(v[1]), int(v[2]))
```

gchanan commented on torch/nn/init.py:
```diff
@@ -525,3 +527,104 @@ def deprecated_init(*args, **kwargs):
+
+    Args:
+        version : Union[Tuple[int, int, int], str], which pytorch version to use for initializing nn.modules.
+            The format should bea version string (major,minor,micro) or a tuple of integers. Examples of valid
```
nit: be a

KushajveerSingh replied:
Done

gchanan commented on torch/nn/init.py:
```diff
@@ -525,3 +527,104 @@ def deprecated_init(*args, **kwargs):
+        version : Union[Tuple[int, int, int], str], which pytorch version to use for initializing nn.modules.
+            The format should bea version string (major,minor,micro) or a tuple of integers. Examples of valid
+            version are (1,7,0), '1.7.1'.
+        use_master : bool, If True, then the latest initialization scheme is used ignoring the value of `version`
```
it's not clear what this means to me.  My interpretation of how this is written is this will go somewhere and download the latest and greatest initialization scheme independent of the version of pytorch that I'm using.  I doubt you meant that though.

Why can't we just pass in torch.__version__ here to get the latest?

KushajveerSingh replied:
Removed `use_master`. It was duplicate functionality.

gchanan commented on torch/nn/init.py:
```diff
@@ -525,3 +527,104 @@ def deprecated_init(*args, **kwargs):
+        2. Now in PyTorch=1.42.0 release a new initialization scheme is introduced.
+        3. To use this new initialization scheme, people would have to change the source code of your repo.
+        This can be a little inconvinient. 
+        4. To avoid this situation you can log a warning during model initialization that our model used this 
```
the warning seems overly prescriptive.  What do you think researchers are concerned about where they would want to warn?

KushajveerSingh replied:
It was suggested by @elistevens 

> The other way around, actually. If I write a model and hardcode 1.7.0 inits into it, then when 2.3.0 comes out with a better init scheme, someone who wants to try applying that new init scheme to my old model has to change the source code.

t-vi replied:
I think the warning is very much wrong as it is currently written, too.

- People wanting to keep the init fixed regardless of PyTorch updates should use a fixed init version.
- Otherwise, one should not.

If anything, I would expect researchers releasing their code to be of the first sort, to facilitate reproducing their results. (To be honest, I would expect most researchers to currently init the things themselves, but that is something else.)
But I also agree with Greg that one might just leave people to choose what they want regardless of the job description.

Also, I would expect that research code basically sets the init somewhere globally and not as much use a context manager.

KushajveerSingh replied:
Removed the warning.

Renamed various things:
* `_init_version` variable -> `init_version`
* `_parse_version` function -> `parse_version`
* `init_version` context manager -> `use_init_version`

> Also, I would expect that research code basically sets the init somewhere globally and not as much use a context manager.

Now, we can do `init.init_version = (1,4,1)`. This just assumes the user passes the input as `Tuple[int, int, int]`. I think we don't have to worry about wrong input here. Added this in the example of `use_init_version`.

KushajveerSingh committed:
Use re to get torch version

KushajveerSingh committed:
Use semantic version syntax to get a common API

_torch_version and _parse_version makes a call to
_get_semantic_version_from_string, making the API much simpler and
easier to debug.

KushajveerSingh committed:
remove use_master

KushajveerSingh committed:
remove use_master from bilinear

KushajveerSingh committed:
remove redundant code from tests

t-vi reviewed (changes requested):
This is a good start, but
- We should have a good plan what we aim at in terms of use. My two users are
  1. Researcher R wants avoid having inits changed when people run her experiments with newer torch because it might change her results.
  2. Anyone else wants the nicest init PyTorch can offer. (And I think the perspective should be this is the "I don't need to do anything".)

This does mean "BC breaking change after deprecation cycle".

Then I think there should be a view how it is typically used.
In particular, do we expect people to use the context manager at all to control initialization (as opposed to a global call/setting)?
If so, I think we should differentiate the overriding from the "end-user" use of the context manager.

KushajveerSingh committed:
Change variable names + rebase master

Renamed various things:
* `_init_version` variable -> `init_version`
* `_parse_version` function -> `parse_version`
* `init_version` context manager -> `use_init_version`

Also, removed the warning in `init_version`.

KushajveerSingh commented on torch/nn/init.py:
```diff
@@ -525,3 +528,97 @@ def deprecated_init(*args, **kwargs):
+
+def _get_semantic_version_from_string(version: str) -> Tuple[int, int, int]:
+    """version: should be 'major.minor.micro'"""
+    v = GET_SEMANTIC_VERSION.findall(version)
```
> Why is this findall?

I have not used `re` much so I used `findall`. Is the below code better?
```
v = GET_SEMANTIC_VERSION.search(version).group(0)
if v is None:
        raise TypeError
```

> Also why is it a a semantic version?

Are you referring to the function name? I changed the function name to `_get_version_as_tuple_from_string`.

KushajveerSingh commented on torch/nn/init.py:
```diff
@@ -525,3 +528,97 @@ def deprecated_init(*args, **kwargs):
+        >>> with nn.init.init_version() as version:
+        >>>     # use the default pytorch initialization used till pytorch version 1.6.1
+        >>>     # This is the default behaviour.
+        >>>     model = nn.Sequential(...)
```
Removed this example. I think it is a redundant example. Also, it might create confusion if we were to change the functionality. \
I think the best use case of context manger is to provide a version to it or set the global.

KushajveerSingh commented on test/test_init_version.py:
```diff
@@ -0,0 +1,63 @@
+        self.assertEqual(x, y)
+
+    def test_version_greater_than_torch(self):
+        # cannot pass version greater than torch.__version__ in init_version
```
If current_torch_version = (1,7,0), then I pass (1,8,0) to `use_init_version`. Now it goes through these checks
1. `use_init_version` raise error if version greater
2. If the above does not raise error, initialize a linear layer to see if it raises an error
3. If the above does not raise error, then raise error that something is wrong.

In the `except` clause I match the error message so that it only checks for the "version greater than" one.

KushajveerSingh commented on torch/nn/init.py:
```diff
@@ -525,3 +528,97 @@ def deprecated_init(*args, **kwargs):
+
+
+@contextlib.contextmanager
+def init_version(version=None):
```
> Why is this only as a context manager?

I think this is how the conversation started. I saw a suggestion to use context manager to solve this issue and I started working from that.

I think this is now resolved by using `init.init_version = (1,7,1)`.

KushajveerSingh committed:
raname _get_semantic_version_from_string

KushajveerSingh commented on torch/nn/modules/linear.py:
```diff
@@ -80,12 +81,18 @@ def __init__(self, in_features: int, out_features: int, bias: bool = True) -> No
-            bound = 1 / math.sqrt(fan_in)
-            init.uniform_(self.bias, -bound, bound)
+    def reset_parameters(self, version: Union[Tuple[int, int, int], str] = None) -> None:
+        with init.init_version(version) as version:
```
> I think this version (an "override version" if you wish) should be handled differently than a user-provided version as the default case would be different.

You mean something like this
```
if version is None:
    # default init ...
else:
    with init.init_version(version) as version:
```

KushajveerSingh replied:
> (Also I think the formatting isn't pep 8 in the type hint of reset_parameters.)

I got the mistake. `Optional[Tuple[int, int, int], str]`. I will update this.

KushajveerSingh commented on test/test_init_version.py:
```diff
@@ -0,0 +1,63 @@
+        self.assertEqual(x, y)
+
+    def test_1_7_0_linear(self):
+        # zero bias is used
```
> How about combining the various versions and have one larger test per layer?

I will add a test function for each modified layer. Will look into the other points you mentioned.

KushajveerSingh commented:
> 1. Researcher R wants avoid having inits changed when people run her experiments with newer torch because it might change her results.

Can this be solved using something as follows?

```python
class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        nn.init.init_version = (1,7,0)
        ...
```

We can go one step further and do `super().__init__(version=(1,6,1))` also.

> In particular, do we expect people to use the context manager at all to control initialization (as opposed to a global call/setting)?

I think the situation is similar to `torch.backends.cudnn.benchmark=True`. Users know what this code line does and they can set this value as per their need.

t-vi commented:
You never want to change a global variable from a module constructor. I think having the user set
```
nn.init.init_version = (1, 7, 0)
```
at the top of the program is reasonable.

KushajveerSingh commented:
@t-vi Made the following changes, can you review these, please:
1. Changed `re.findall` to `re.match`.
2. Changed `Union[Tuple, str]` to `Union[Tuple, str, None]` to follow PEP 8.
3. Updated the doc of `use_init_version` to tell that the default behaviour (`version=None`) will change in the future.
4. Added initialization documentation in `nn.Linear` (if no changes need to be made, I will add this to other modules).
5. Added `use_init_version` in `nn.init` documentation page.
6. Modified `test_version_greater_than_torch` to use `assertRaises`.

@t-vi Can you expand on this?
> I think this version (an "override version" if you wish) should be handled differently than a user-provided version as the default case would be different.

Also, can you give some pointers for this I am lost?
> To avoid flakiness, we would want to add some seeding thing (I think other tests do this, but I don't know if its done for all by default), but then we should check for the mean and std being in some reasonable CI from what is expected.

![linear](https://user-images.githubusercontent.com/24699564/89054384-555d4900-d376-11ea-932f-f5262431b8c5.png)

![init](https://user-images.githubusercontent.com/24699564/89055074-71151f00-d377-11ea-9424-b739be47c6a8.png)

KushajveerSingh committed:
Make changes as suggested in code review

1. Changed re.findall to re.match.
2. Changed Union[Tuple, str] to Union[Tuple, str, None] to follow PEP 8.
3. Updated the doc of use_init_version to tell that the default behaviour (version=None) will change in the future.
4. Added initialization documentation in nn.Linear (if no changes need to be made, I will add this to other modules).
5. Added use_init_version in nn.init documentation page.
6. Modified test_version_greater_than_torch to use assertRaises.

KushajveerSingh committed:
solve mypy issue

t-vi commented:
If we want to do this we would need to refresh.
I don't think passing the version to reset_parameters is a good idea, we should drop it (per "there should be preferably only one way to do it"). If the user has lots of opinions on versions, they should write their own init.

I don't think using the context manager in the init is actually needed.

Also, we should not let the user manipulate the version without checking, so assigning the global variable doesn't look too good to me.

@KushajveerSingh I realise that this has been sitting with "review requested from t-vi" for along time, but are you still interested in this?

KushajveerSingh commented:
Is this useful? This is a tricky question to answer. The main argument that I am considering is initialization schemes have become more architecture dependent. But still there are many cases where I am writing the same initialization code for some layers.

From a user perspective, I want something simple like I declare/specify somehow that I want to use Kaiming init for all Linear layers in my project. I don't want to learn some new syntax to make this happen.

What are some of the possible solutions (that give flexibility to Pytorch to add new inits without having to break things)? I don't know. I agree context manager is not needed.

@t-vi What are your thoughts about this? If we create a central class that kind of hooks into the initialization of all layers. A user can provide arguments to this class like `torch.manual_seed(23)` on what initialization to use for each layer.

albertz commented:
Hm, GitHub does not let me add comments to the changed code directly?

Anyway, I just wanted to say, I think the `mode='fan_out'` is wrong in many places, and it should be `mode='fan_in'`. Also see my comment here: https://github.com/pytorch/pytorch/issues/18182#issuecomment-1016433802

t-vi commented:
It doesn't look to me like this would fully work, so I'd be inclined to close it, even if I still think that this is one of PyTorch's murkier bits. @KushajveerSingh what do you think?
